src/sonic-py-common/sonic_py_common/multi_asic.py

```diff
@@ -0,0 +1,290 @@
+                back_ns.append(namespace)
+
+    return {'front_ns': front_ns, 'back_ns': back_ns}
+
+
+def get_namespace_list(namespace=None):
+    if not is_multi_asic():
+        ns_list = [DEFAULT_NAMESPACE]
+        return ns_list
+
+    if namespace is None:
+        namespaces = get_all_namespaces()
+        ns_list = namespaces['front_ns'] + namespaces['back_ns']
+    else:
+        ns_list = [namespace]
+
+    return ns_list
+
+
+def get_port_table(namespace=None):
+    """
+    Retrieves the ports from all the asic present on the devices
+
+    Returns:
+        a dict of all the ports
+    """
+    all_ports = {}
+    ns_list = get_namespace_list(namespace)
+
+    for ns in ns_list:
+        ports = get_port_table_for_asic(ns)
+        all_ports.update(ports)
+
+    return all_ports
+
+
+def get_port_table_for_asic(namespace):
+
+    config_db = connect_config_db_for_ns(namespace)
+    ports = config_db.get_table(PORT_CFG_DB_TABLE)
+    return ports
+
+
+def get_namespace_for_port(port_name):
+
+    ns_list = get_namespace_list()
+    port_namespace = None
+
+    for ns in ns_list:
+        ports = get_port_table_for_asic(ns)
+        if port_name in ports:
+            port_namespace = ns
+            break
+
+    if port_namespace is None:
+        raise ValueError('Unknown port name {}'.format(port_name))
+
+    return port_namespace
+
+
+def get_port_role(port_name, namespace=None):
+
+    ports_config = get_port_table(namespace)
+    if not ports_config[port_name].has_key('role'):
+        return EXTERNAL_PORT
+
+    role = ports_config[port_name]['role']
+    return role
+
+
+def is_port_internal(port_name, namespace=None):
+
+    role = get_port_role(port_name, namespace)
+
+    if role is INTERNAL_PORT:
+        return True
+
+    return False
+
+
+def is_port_channel_internal(port_channel, namespace=None):
+
+    if not is_multi_asic():
+        return False
+
+    ns_list = get_namespace_list(namespace)
+
+    for ns in ns_list:
+        config_db = connect_config_db_for_ns(ns)
+        port_channels = config_db.get_table(PORT_CHANNEL_CFG_DB_TABLE)
+
+        if port_channel in port_channels:
+            members = port_channels[port_channel]['members']
+            if is_port_internal(members[0], namespace):
+                return True
+
+    return False
+
+
+def is_bgp_session_internal(bgp_neigh_ip, namespace=None):
+
+    if not is_multi_asic():
+        return False
+
+    ns_list = get_namespace_list(namespace)
+
+    for ns in ns_list:
+
+        config_db = connect_config_db_for_ns(ns)
+        bgp_sessions = config_db.get_table(BGP_NEIGH_CFG_DB_TABLE)
+        if bgp_neigh_ip not in bgp_sessions:
+            continue
+
+        bgp_neigh_name = bgp_sessions[bgp_neigh_ip]['name']
+        neighbor_metadata = config_db.get_table(
+            NEIGH_DEVICE_METADATA_CFG_DB_TABLE)
+
+        if neighbor_metadata[bgp_neigh_name]['type'].lower() == 'asic':
```

Comment:
do we need to check if neighbor_metadata exist ?

Reply:
added a check in latest commit